components/hab/src/cli.rs

```diff
@@ -1298,9 +1305,22 @@ pub fn sub_sup_bash() -> App<'static, 'static> {
     )
 }
 
+fn config_file_to_defaults(config_file: &str)
+                           -> Result<PartialSubSupRun, Box<dyn std::error::Error>> {
+    let contents = fs::read_to_string(config_file)?;
+    Ok(toml::from_str(&contents)?)
+}
+
 pub fn sub_sup_run(feature_flags: FeatureFlag) -> App<'static, 'static> {
     let sub = if feature_flags.contains(FeatureFlag::CONFIG_FILE) {
-        SubSupRun::clap()
+        let mut sub = SubSupRun::clap();
+        if let Ok(config_file) = env::var("HAB_FEAT_CONFIG_FILE") {
+            match config_file_to_defaults(&config_file) {
+                Ok(defaults) => configopt::set_defaults(&mut sub, &defaults),
```

Comment:
This flow seems confusing... it seems like you're determining the CLI arguments first, and then grabbing configuration from the config file (if it exists) to... do something with default values underneath that?

I would have expected configuration file values to be read in first, and then overlaid with any CLI and / or environment overrides.

Am I overlooking something here?

Reply:
Agreed, this is confusing. Actually, what you describe is exactly what is happening, "configuration file values to be read in first, and then overlaid with any CLI". What is confusing is how the overlay happens. The overlay happens by setting/overriding the defaults of the CLI args. So to go line by line:

- 1316: Creates a `clap::App` from the `SubSupRun` definition using `structopt`
- 1318: Reads the config file and converts it to a `PartialSubSupRun` which implements `ConfigOptDefaults`. `ConfigOptDefaults` is a trait that can be used to override the default values of a `clap::App`
- 1319: Set/override the cli defaults with values from the config file
- We are now ready to parse the CLI input taking into account the newly set default values from the config file

The reason we need the song and dance of overriding the default values of the CLI (opposed to merging the CLI and config file after the parsing the CLI input) is to allow the user to not set values on the command line if they are set in the config file. If we did not add a default value parsing the CLI input would fail.

Once we switch to using `structopt` to parse the CLI opposed to `clap`'s `matches.vaue_of("ARG_NAME")` this logic will not be necessary.

Reply:
If it changes out the defaults while still allowing user-specified data to show through (and doesn't replace user-specified data with defaults!), that's cool. The fact that this goes away is also nice.

Adding some comments that clarify this situation would also be nice.